src/Common/Auth/AuthUtils.php

```diff
@@ -943,35 +1005,90 @@ private function checkPasswordNotExpired($user)
     {
-        if ($GLOBALS['password_max_failed_logins'] == 0 || self::useActiveDirectory($user)) {
-            // skip the check if turned off or using active directory for login
+        if ($GLOBALS['password_max_failed_logins'] == 0) {
+            // skip the check if turned off
             return true;
         }
 
-        $query = privQuery("SELECT `login_fail_counter` FROM `users_secure` WHERE BINARY `username` = ?", [$user]);
+        $query = privQuery("SELECT `login_fail_counter`, TIMESTAMPDIFF(SECOND, `last_login_fail`, NOW()) as `seconds_last_login_fail` FROM `users_secure` WHERE BINARY `username` = ?", [$user]);
         if ($query['login_fail_counter'] >= $GLOBALS['password_max_failed_logins']) {
+            if (!empty($GLOBALS['time_reset_password_max_failed_logins']) &&
+                $GLOBALS['time_reset_password_max_failed_logins'] > 0 &&
+                !empty($query['seconds_last_login_fail']) &&
+                $query['seconds_last_login_fail'] > $GLOBALS['time_reset_password_max_failed_logins']
+            ) {
+                // the last login fail was longer than the timeout required to reset the failed logins, so will pass
+                //  (also need to reset the counter)
+                self::resetLoginFailedCounter($user);
+                return true;
+            }
             return false;
         } else {
             return true;
         }
     }
 
-    private function resetLoginFailedCounter($user)
+    private function checkIpLoginFailedCounter(string $ipString): array
     {
-        if (!self::useActiveDirectory($user)) {
-            // skip if using active directory for login
-            privStatement("UPDATE `users_secure` SET `login_fail_counter` = 0 WHERE BINARY `username` = ?", [$user]);
+        if ($GLOBALS['ip_max_failed_logins'] == 0) {
+            // skip the check if turned off
+            return ['pass' => true, 'force_block' => null, 'skip_timing_attack' => null];
+        }
+
+        $query = privQuery("SELECT `ip_force_block`, `ip_no_prevent_timing_attack`, `ip_login_fail_counter`, TIMESTAMPDIFF(SECOND, `ip_last_login_fail`, NOW()) as `seconds_last_ip_login_fail` FROM `ip_tracking` WHERE `ip_string` = ?", [$ipString]);
+        if ($query['ip_force_block'] == 1) {
+            if ($query['ip_no_prevent_timing_attack'] == 1) {
+                return ['pass' => false, 'force_block' => true, 'skip_timing_attack' => true];
+            } else {
+                return ['pass' => false, 'force_block' => true, 'skip_timing_attack' => false];
+            }
+        }
+        if ($query['ip_login_fail_counter'] >= $GLOBALS['ip_max_failed_logins']) {
+            if (!empty($GLOBALS['ip_time_reset_password_max_failed_logins']) &&
+                $GLOBALS['ip_time_reset_password_max_failed_logins'] > 0 &&
+                !empty($query['seconds_last_ip_login_fail']) &&
+                $query['seconds_last_ip_login_fail'] > $GLOBALS['ip_time_reset_password_max_failed_logins']
```

Comment:
seems like you could end up with some wierd bugs if someone were to put a letter in the globals, would it be better to grab an intval of $GLOBALS['ip_time_reset_password_max_failed_logins'] and do the comparison that way?

Reply:
yep, prob good idea to go through the globals that using in this pr and ensure are int
TODO for me